redgenes/bakta_annotations.py

```diff
@@ -0,0 +1,189 @@
+    dtype_map = {
+        "taxid": "Int64",
+        "species_taxid": "Int64",
+        "gc_percent": float,
+        "replicon_count": "Int64",
+        "scaffold_count": "Int64",
+        "contig_count": "Int64",
+        "total_gene_count": "Int64",
+        "protein_coding_gene_count": "Int64",
+        "non_coding_gene_count": "Int64",
+    }
+    df = df.astype(dtype_map)
+    return df
+
+def extract_bakta_results(gff_path: Path):
+    """
+    Extract information from a Bakta GFF output file.
+    """
+    # Use the determined position to read the file, skipping rows up to the header
+    column_names = ["contig_ID", "type", "start", "stop", "strand", "locus_tag", "gene", "product", "dbxrefs"]
+    df = pd.read_csv(gff_path, header=0, names=column_names, comment='#', sep='\t')
+    return df
+
+def annotation_pipeline(df: pd.DataFrame, tmpdir: Path, bakta_db: Path):
+    """
+    Annotate genomes based on Bakta and insert information into the database.
+    """
+    for _, row in df.iterrows():
+        local_path, filename = Path((row["local_path"])), row["assembly_accession"]
+        source, external_accession = row["source"], row["assembly_accession"]
+        with copy_and_unzip(str(local_path), tmpdir) as input_fasta:
+            bakta_output_tsv, bakta_commands = run_bakta(input_fasta, tmpdir, bakta_db)
+        bakta_df = extract_bakta_results(str(bakta_output_tsv))
+
+        with TRN:
+            sql_identifier = """
+                INSERT INTO identifier (filename_full, filepath)
+                VALUES (?, ?)
+                RETURNING entity_id"""
+            args_identifer = [filename, str(local_path)]
+            TRN.add(sql_identifier, args_identifer)
+            entity_id = TRN.execute_fetchflatten()
+
+            sql_md_info = """
+                INSERT INTO md_info (entity_id, source, external_accession)
+                VALUES (?, ?, ?)"""
+            args_md_info = [source, external_accession]
+            args_md_info = entity_id + args_md_info
+            TRN.add(sql_md_info, args_md_info)
+
+            sql_extract_bakta_run_info = """
+                SELECT run_id FROM run_info
+                WHERE software = 'bakta' and version = ? and commands = ?"""
+            TRN.add(
+                sql_extract_bakta_run_info, [bakta_version, bakta_commands]
+            )
+            bakta_run_id = TRN.execute_fetchflatten()
+
+            if not bakta_run_id:
+                sql_run_info_bakta = """
+                    INSERT INTO run_info (software, version, commands)
+                    VALUES (?, ?, ?) RETURNING run_id"""
+                args_run_info_bakta = [
+                    "bakta",
+                    bakta_version,
+                    bakta_commands,
+                ]
+                TRN.add(sql_run_info_bakta, args_run_info_bakta)
+                bakta_run_id = TRN.execute_fetchflatten()
+
+            bakta_df.insert(0, "entity_id", entity_id[0])
```

Comment:
I think this is the same as `bakta_df['entity_id'] = entity_id[0]`, right? 

Columns can be re-ordered subsequently with `bakta_df[list_of_columns_in_desired_order]`